### PR TITLE
Fixe DoG accuracy, add `upscale_double`

### PR DESCRIPTION
**Pull request by vicsyl**
#### Changes
Fixes bugs in DoG detector. Changes the upsampling to upscale_double, which is rotation/reflection equivariant.  

#### Type of change
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: kornia/feature/scale_space_detector.py [hunk 1]
```diff
@@ -203,17 +204,18 @@ def detect(self, img: Tensor, num_feats: int, mask: Optional[Tensor] = None) ->
             resp_flat_best = resp_flat_best * good_mask.to(dev, dtype)
 
             # Normalize LAFs
-            current_lafs = normalize_laf(current_lafs, octave[:, 0])  # We don`t need # of scale levels, only shape
+            current_lafs *= px_size
```
Comment (edgarriba) [inline kornia/feature/scale_space_detector.py:207]: in place won't work with autograd

File: kornia/geometry/subpix/spatial_soft_argmax.py [hunk 2]
```diff
@@ -632,6 +633,8 @@ def conv_quad_interp3d(input: Tensor, strict_maxima_bonus: float = 10.0, eps: fl
         y_max += strict_maxima_bonus * new_nms_mask.to(input.dtype)
 
     dx_res: Tensor = dx.flip(1).reshape(B, CH, D, H, W, 3).permute(0, 1, 5, 2, 3, 4)
+    dx_res[:, :, [1, 2]] = dx_res[:, :, [2, 1]]
```
Comment (edgarriba) [inline kornia/geometry/subpix/spatial_soft_argmax.py:636]:
```suggestion
    dx_res[:, :, (1, 2)] = dx_res[:, :, (2, 1)]
```
tuples are immutable

File: kornia/geometry/transform/pyramid.py [hunk 3]
```diff
@@ -410,3 +409,24 @@ def build_laplacian_pyramid(
...
+    r"""Upscale image by the factor of 2, even indices maps to original indices.
+
+    Odd indices are linearly interpolated from the even ones.
+
+    Args:
+        x : input image.
+    Shape:
+        - Input: :math:`(B, C, H, W)`
+        - Output :math:`(B, C, H, W)`
+    """
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:424]: need to add checkers `KORNIA_CHECK_SHAPE(x, ["*", "H", "W"])`

and i would adjust the function to work with arbitrary 2d shape

File: test/integration/test_dog.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
+import math
```
Comment (edgarriba) [inline test/integration/test_dog.py:1]: i'd add this somewhere in https://github.com/kornia/kornia-benchmark

we are trying to reduce the size of the repository to keep as light as possible, including heavy tests like this.

in case we want to include as part of our ci, ideally it must be only torch. I see that big part of the opencv functionality you use is found already here in kornia.

/cc @ducha-aiki @johnnv1

File: kornia/feature/scale_space_detector.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/feature/scale_space_detector.py:207]: @edgarriba this kind of inplace works usually

**Comment by ducha-aiki** [comment]
@edgarriba that is a work I mentioned to you. I will review it soon.

**Comment by ducha-aiki** [comment]
It seems that we also have to fix a bit `normalize_laf` and `denormalize_laf`

File: kornia/feature/scale_space_detector.py (hunk 1 above)
Comment (edgarriba) [reply kornia/feature/scale_space_detector.py:207]: oh, good to know !!

File: kornia/geometry/subpix/spatial_soft_argmax.py [hunk 5]
```diff
@@ -620,7 +620,8 @@ def conv_quad_interp3d(input: Tensor, strict_maxima_bonus: float = 10.0, eps: fl
     #  Kill those points, where we cannot solve
     new_nms_mask = nms_mask.masked_scatter(nms_mask, solved_correctly)
 
-    x_solved.masked_scatter_(new_nms_mask.view(-1, 1, 1), x_solved_masked[solved_correctly])
+    x_solved[torch.where(new_nms_mask.view(-1, 1, 1))[0]] = x_solved_masked[solved_correctly]
```
Comment (johnnv1) [inline kornia/geometry/subpix/spatial_soft_argmax.py:623]:
```suggestion
    x_solved[where(new_nms_mask.view(-1, 1, 1))[0]] = x_solved_masked[solved_correctly]
```

from kornia.core import where

File: test/integration/test_dog.py (hunk 4 above)
Comment (johnnv1) [reply test/integration/test_dog.py:1]: I also don't see why we would have this test here. If not in kornia-benchmark maybe add it as part of the tutorials?  Also the images could be included in the kornia/data repository

File: kornia/geometry/transform/pyramid.py [hunk 6]
```diff
@@ -410,3 +409,24 @@ def build_laplacian_pyramid(
         laplacian_pyramid.append(laplacian)
     laplacian_pyramid.append(gaussian_pyramid[-1])
     return laplacian_pyramid
+
+
+def upscale_double(x):
```
Comment (johnnv1) [inline kornia/geometry/transform/pyramid.py:414]:
```suggestion
def upscale_double(x: Tensor) -> Tensor:
```

File: kornia/geometry/transform/pyramid.py [hunk 7]
```diff
@@ -410,3 +409,24 @@ def build_laplacian_pyramid(
...
+
+    Args:
+        x : input image.
+    Shape:
+        - Input: :math:`(B, C, H, W)`
+        - Output :math:`(B, C, H, W)`
+    """
+    B, CH, H, W = x.shape
+    upscaled = torch.zeros((B, CH, H * 2, W * 2))
+    upscaled[:, :, ::2, ::2] = x
```
Comment (johnnv1) [inline kornia/geometry/transform/pyramid.py:427]: this isnt an option here?
```suggestion
    upscaled[..., ::2, ::2] = x
```

Comment (edgarriba) [reply kornia/geometry/transform/pyramid.py:427]: +1 for (*,H,W) unless there's objection

Comment (vicsyl) [reply kornia/geometry/transform/pyramid.py]: implemented

File: kornia/geometry/transform/pyramid.py (hunk 6 above)
Comment (vicsyl) [reply kornia/geometry/transform/pyramid.py]: done

File: kornia/geometry/transform/pyramid.py (hunk 3 above)
Comment (vicsyl) [reply kornia/geometry/transform/pyramid.py:433]: done

File: kornia/geometry/subpix/spatial_soft_argmax.py (hunk 2 above)
Comment (vicsyl) [reply kornia/geometry/subpix/spatial_soft_argmax.py]: done

File: test/integration/test_dog.py (hunk 4 above)
Comment (vicsyl) [reply test/integration/test_dog.py]: I shortened it to absolute minimum - to 100 lines. It uses torch/kornia only except for a) cv.imread/cvtColor and b)  np.rot90 to prepare imgs for logic based on cv (in another file, another 100 lines), which gives better results (reprojection error ~0.001 vs ~0.1), I suspect because of stability of kornia.geometry.RANSAC vs. cv.findHomography.
 
Maybe it is still heavy-weight - I can move it to kornia-benchmark, I just wanted to show what this change does (ensures/comes very close to rotational equivariance of DoG) - I guess in kornia-benchmark it will only be possible to see the improvement once this code is released.

File: kornia/geometry/subpix/spatial_soft_argmax.py (hunk 5 above)
Comment (vicsyl) [reply kornia/geometry/subpix/spatial_soft_argmax.py]: done

File: test/geometry/transform/test_pyramid.py [hunk 8]
```diff
@@ -192,3 +192,31 @@ def test_gradcheck(self, device, dtype):
...
         )
+
+
+class TestUpscaleDouble:
+    @pytest.mark.parametrize("shape", ((10, 10), (3, 10, 10), (1, 3, 10, 10)))
+    def test_upscale_double(self, shape, device, dtype):
+        print(f"shape: {shape}")
+        print(f"device: {device}")
+        print(f"dtype: {dtype}")
+
```
Comment (johnnv1) [inline test/geometry/transform/test_pyramid.py:203]:
```suggestion

```

File: test/geometry/transform/test_pyramid.py [hunk 9]
```diff
@@ -192,3 +192,31 @@ def test_gradcheck(self, device, dtype):
             raise_exception=True,
             fast_mode=True,
         )
+
+
+class TestUpscaleDouble:
```
Comment (johnnv1) [inline test/geometry/transform/test_pyramid.py:197]:
```suggestion
class TestUpscaleDouble(BaseTester):
```
`from kornia.testing import BaseTester`

File: test/geometry/transform/test_pyramid.py [hunk 10]
```diff
@@ -192,3 +192,31 @@ def test_gradcheck(self, device, dtype):
...
+            x = xm
+        else:
+            x[..., 0, :, :] = xm
+            x[..., 1, :, :] = ym
+
+        upscaled = kornia.geometry.transform.upscale_double(x)
+        assert torch.all(
+            upscaled[..., 1::2, 1::2][..., :-1, :-1]
+            == (upscaled[..., ::2, ::2][..., :-1, :-1] + upscaled[..., ::2, ::2][..., 1:, 1:]) / 2
+        )
```
Comment (johnnv1) [inline test/geometry/transform/test_pyramid.py:220]: add a hardcoded expected value instead of this one

File: test/geometry/transform/test_pyramid.py [hunk 11]
```diff
@@ -192,3 +192,31 @@ def test_gradcheck(self, device, dtype):
...
+        downscaled_back = upscaled[..., ::2, ::2]
+        assert torch.all(x == downscaled_back)
```
Comment (johnnv1) [inline test/geometry/transform/test_pyramid.py:222]:
```suggestion
        assert_close(x, downscaled_back)
```

`from kornia.testing import assert_close`

File: test/integration/test_dog.py (hunk 4 above)
Comment (ducha-aiki) [reply test/integration/test_dog.py]: @vicsyl there should be no `cv2` in the main repo. How about the following:
1) put the smallest possible image size into `torch.save` file, similarly how we test LoFTR here: https://github.com/kornia/kornia/blob/master/test/feature/test_loftr.py#L37

The rest of benchmark can go to the kornia-benchmark and there is no restrictions on the imports there.

**Review by ducha-aiki — APPROVED** [review]
LGTM. One thing: if `normalize_laf` is bugged, it would make sense to ti fix them instead of getting rid of them.

File: test/geometry/transform/test_pyramid.py [hunk 12]
```diff
@@ -192,3 +192,100 @@ def test_gradcheck(self, device, dtype):
...
+            expected = expected[None]
+
+        assert_close(upscaled, expected)
+        assert torch.all(upscaled == expected)
+        downscaled_back = upscaled[..., ::2, ::2]
+        assert torch.all(x == downscaled_back)
+
+    @staticmethod
+    def prepare_data(shape, device, dtype, requires_grad=False):
+        xm = torch.tensor([[0, 0, 0, 0, 0], [1, 1, 1, 1, 1], [2, 2, 2, 2, 2], [3, 3, 3, 3, 3], [4, 4, 4, 4, 4]]).to(
```
Comment (edgarriba) [inline test/geometry/transform/test_pyramid.py:276]:
```suggestion
        xm = torch.tensor([[0, 0, 0, 0, 0], [1, 1, 1, 1, 1], [2, 2, 2, 2, 2], [3, 3, 3, 3, 3], [4, 4, 4, 4, 4]], device=device, dtype=dtype)
```
otherwise cuda tests will fail. Not relevant at all for tests, but is faster to pass to constructor during tensor allocation

File: kornia/geometry/transform/pyramid.py [hunk 13]
```diff
@@ -410,3 +418,26 @@ def build_laplacian_pyramid(
         laplacian_pyramid.append(laplacian)
     laplacian_pyramid.append(gaussian_pyramid[-1])
     return laplacian_pyramid
+
+
+def upscale_double(x: Tensor) -> Tensor:
+    r"""Upscale image by the factor of 2, even indices maps to original indices.
```
Comment (edgarriba) [inline kornia/geometry/transform/pyramid.py:424]: add function here to the docs: https://github.com/kornia/kornia/blob/master/docs/source/geometry.transform.rst

**Review by edgarriba — APPROVED** [review]
few nitpicks but LGTM overall
